### ISSUE TITLE
port twisted.names.authority to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8259]
|[<img alt="pawelmhm's avatar" src="https://avatars.githubusercontent.com/u/2700942?s=50" width="50" height="50">](https://github.com/pawelmhm)| @pawelmhm reported|
|-|-|
|Trac ID|trac#8259|
|Type|enhancement|
|Created|2016-03-28 07:51:27Z|
|Branch|https://github.com/twisted/twisted/tree/t-names-authority-py3-8259-3|

See previous ticket [#8195](https://github.com/twisted/twisted/issues/8195) (ticket)

Attachments:

* [8259.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1d5/1d57fe248fb3a3c8dc81529a2188091471251a1e/5ed2ddbc6805901c1450f6d13eb6698b953d9018.patch) (16387 bytes) - added by pawelmhm on 2016-03-29 18:44:41Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__8259 8259
type__enhancement enhancement
reporter__pawelmhm pawelmhm
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_t_names_authority_py3_8259_3 branches/t-names-authority-py3-8259-3
branch_author__hawkowl__adiroiban hawkowl, adiroiban
status__closed closed
resolution__fixed fixed
component__names names
keywords__None None
time__1459151487524737 1459151487524737
changetime__1460631153367253 1460631153367253
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="pawelmhm's avatar" src="https://avatars.githubusercontent.com/u/2700942?s=50" width="50" height="50">](https://github.com/pawelmhm)|@pawelmhm set owner to @pawelmhm|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="pawelmhm's avatar" src="https://avatars.githubusercontent.com/u/2700942?s=50" width="50" height="50">](https://github.com/pawelmhm)<a name="note_2"></a>|@pawelmhm removed owner|
|-|-|

Here's the patch. It ports twisted.names.authority and twisted.names.secondary + it ports twisted.test.test_names tests and twisted.names.test_examples tests. 

Fixes:

* usual trivial Python 3 porting stuff, e.g. replacing StringIO with NativeStringIO, removing deprecated syntax where it matters 
* ip addresses should be string (not bytes) because this is what socket.inet_aton accepts on Python 3 (it was accepting bytes on Python 2)
* domain names should be bytes
* DNS records inheriting from TXT record need to accept bytes as some arguments to init

let me know if you find any issues

**Comment by twisted-trac** [comment]
|[<img alt="pawelmhm's avatar" src="https://avatars.githubusercontent.com/u/2700942?s=50" width="50" height="50">](https://github.com/pawelmhm)|@pawelmhm set owner to @adiroiban|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

(In [47122]) Branching to t-names-authority-py3-8259.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban removed owner|
|-|-|

It looks good

some pyflakes and twistedchecker error

```
************* Module twisted.names.test.test_names
C0301:352,0: Line too long (91/79)
C0301:363,0: Line too long (100/79)

twisted/names/test/test_names.py:12: 'NativeStringIO' imported but unused
```

-------

branch created for branch and patch sent to buildbot.

I don't have time now to investigate if we should have bytes or unicode...but please use explicit text marker that is use `u'text'` or `b'text'` when changing types

Leaving this for review so that another persons should check it

I would argue that the API should also accept bytes... but I have never used twisted.named and I don't have a good understanding of DNS APIs.

Thanks for the patch!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl set owner to @hawkowl|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl commented|
|-|-|

Hi, thanks for this patch, pawelmhm.

The bytes/Unicode separation issues are a bit complex; I agree with Adi's "also accepting bytes", and I'm also wary of the fact that Twisted handles IP addresses and URLs as "text" (i.e. unicode) in most parts of Twisted on Python 3. Since a lot of this is tiny changes and a lot of finicky stuff, writing the review so you could change it would take more time for me to write and also waste your time, so I went ahead and built off your changes.

After reading through the code, and thinking on it, I think that handling bytes and Unicode strings is important wherever we handle (just) URLs or IP addresses, and so we should accept Unicode and bytes on both platforms, encoding using idna or ascii where required (as both URLs and IP addresses should be ASCII). As such, I've made the tests use a mixture of declared bytes (what it serves) and undeclared str (so, Unicode on Python 3) in the comparisons, to ensure that both are handled correctly and that they serialise down to the same value. To this end, I've done a lot of little docstrings that say that they accept bytes or unicode -- but only in the ```__init__``` - the instance variables are all always bytes, in a binary packed format (so, something you send over the wire, an encoded version of what you give to the init).

The patch has grown quite big as a result of lots of little line changes along the way (e.g. autoformatting by python-docstring-mode), so I will whittle it down so it's reviewable.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl commented|
|-|-|

(In [47188]) Branching to t-names-authority-py3-8259-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl removed owner|
|-|-|

I'm happy with this. Since I did a lot of changes (mostly adding docstrings), I'll put it up for review. This patch should catch all of the ambiguous types taken by it, but leaves in a lot of bare strs in the tests so that handling those is handled too.

[#8279](https://github.com/twisted/twisted/issues/8279) is the cause of the two Fedora 23 2.7 builders failing, so you can disregard those.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph set owner to @hawkowl|
|-|-|

Looks good; lots of new docstrings, `ResourceWarning` fixes, type fixes.  Thanks.

Two minor issues you might want to fix before landing:

1. You left a few `C{bytes}` in various places where you changed `str` to `bytes`; particularly jarring when it's now `C{bytes} or L{unicode}`.
2. Maybe put a comment in as to why you're using `NativeStringIO` for stdin/stdout in tests?  If find the `ioType` of `sys.stdout` sort of baffling so some redundancy there never hurts :).
3. Since [#8279](https://github.com/twisted/twisted/issues/8279) is fixed, maybe merge forward just to get a pre-commit full test run just to be sure.

Otherwise, please land!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl commented|
|-|-|

(In [47223]) Branching to t-names-authority-py3-8259-3.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl set status to `closed`|
|-|-|

(In [47227]) Merge t-names-authority-py3-8259-3: Port twisted.names.authority to Python 3

Author: pawelmhm, hawkowl
Reviewers: adiroiban, hawkowl, glyph
Fixes: [#8259](https://github.com/twisted/twisted/issues/8259)
